pkg/controller/clone-controller.go

```diff
@@ -226,7 +227,14 @@ func (cc *CloneController) Run(threadiness int, stopCh <-chan struct{}) error {
 	return nil
 }
 
-func (cc *CloneController) runPVCWorkers() {
+//RunWorkers is declared in Controller WorkerRunner interface
+func (cc *CloneController) RunWorkers() {
```

Comment:
Just an idea for further consolidation; we could probably move these into Controller.go and operate on the Controller struct itself instead of the individual import, upload and clone controllers.  As it stands now we have an exported Run, ProcessNextPvcItem etc duplicated in each controller, that can be cleaned up pretty easily.  We could put these in the Controller.go pkg and just use a Controller as the receiver.  In the case of unique things that are still left we could just use the ctrlType identifier to call an unexported helper function in the appropriate controller pkg.  One problem with this is we're kinda reveresed on how I think we should do the 'has-a' relationship.  Rather than the import, clone or upload controllers having a base Controller, we should have everything be a Controller, and it include an instance of the specific controller we want.  If we do that then a lot of the logic falls into place with these and as an added bonus from an exported process flow perspective we only have to trace the flow through a single set of Controller->XYZ funcs.  It gives us a nice pattern and a sort  of contract for our controllers.

I'm not proposing you rework this patch to add that sort of thing, but I think if you're interested it would be a good next step in the consolidation/cleanup process.  I'd be happy to share some ideas and collaborate with your or anyone else that is interested.

Reply:
ProcessNextPvcItem in upload controller does different things a little bit than the other controllers so this is the reason I left it like this. 
Run is being called from cmd/cdi-controller/controller.go and I need to see if we can move it to the (sub class) controller. did not think about it.
What do you think about merging it as it now (still there was lots of work here refactoring the code and I'm not sure you can change ProcessNextPvcItem in the uploader to be like the others, need to see) and tackle it as you said in another cleanup  pr or something..... Will the pvc transfer controller goes into CDI as well? that will be another chance. Let me know.... In the meantime, I will push the other change (ctlType vars).